feature: add webhooks customizable from admin panel

indyteo opened the pull request:
**Issue:** #892 / #893 / [Webhook updates on events](https://feedback.fider.io/posts/185/webhook-updates-on-events) / [Add Slack Integration](https://feedback.fider.io/posts/36/add-slack-integration)

Please check this discussion (https://github.com/getfider/fider/discussions/969#discussioncomment-1013585) to see every details.

Note that I based the general idea on what @acamilleri did in his PR and extended it from the environment variable to the admin control panel.

goenning commented:
I started reviewing this today, didn't went through code yet, but I love how flexible the templates are and the feature is working end to end. Great work!

My suggestions below, I would like to hear your opinion if you don't agree with some of these :)

1. HTTP Method could be a dropdown, GET/POST/PUT/DELETE/PATCH should be enough?!
2. Header Name/Value might be better in the same row
3. I was able to save a webhook with invalid content: eg: `{{ boom .post_title }}`. Is it possible to validate before saving?
4. Some tooltips have a "hand" icon but no click action. We could use the "?" or pointer cursor instead, only use the hand if there's a click action.
5. Preview looks great! It's super useful!

goenning commented:
btw, I had to change the migration filename to make it work

indyteo commented:
Thanks for starting the review!

About your questions:
1. I first planed to do this, but those methods are just the standardized ones, and a server could implement custom values: [See RFC-7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4). Maybe a drop down with common values and a special "Other" value that show an additional text input field allowing a custom value.
2. What do you mean by "in the same row"? Isn't it the case now? Could you please join a screenshot explaining your expectation about this? Thanks
3. Hmm, it seems possible. We would need to parse the template in the validation, an check if an error occurred. We can consider this option.
4. I totally agree, but this is the tooltip style that define the hand cursor. I don't know why it was defined like that but if you think it was a mistake we can fix it.
5. I don't think a huge modal is a problem, in fact the length of the content won't change as we have the same things to display. But the modal allow users to scroll the help separately from the rest of the page. I think we're in fact using the best visualization method.
6. Yes, I'm pretty proud of it 😊

goenning commented:
> 1. I first planed to do this, but those methods are just the standardized ones, and a server could implement custom values: [See RFC-7231](https://datatracker.ietf.org/doc/html/rfc7231#section-4). Maybe a drop down with common values and a special "Other" value that show an additional text input field allowing a custom value.

Fair point, but having a dropdown and an input would just look more complicated, so let's keep it simple.

> 2. What do you mean by "in the same row"? Isn't it the case now? Could you please join a screenshot explaining your expectation about this? Thanks

![image](https://user-images.githubusercontent.com/94755/127742756-00536927-90ef-4574-985a-b509802db0b7.png)

I think header and value could be side by side, possibly a small header input and a larger value input.

> 3. Hmm, it seems possible. We would need to parse the template in the validation, an check if an error occurred. We can consider this option.

It's kind of like running the preview as a validation for it.

> 4. I totally agree, but this is the tooltip style that define the hand cursor. I don't know why it was defined like that but if you think it was a mistake we can fix it.

Ok, I see now. That's a mistake, all the existing tooltips are not clickable, so the "pointer" cursor doesn't seem like the correct one.

> 5. I don't think a huge modal is a problem, in fact the length of the content won't change as we have the same things to display. But the modal allow users to scroll the help separately from the rest of the page. I think we're in fact using the best visualization method.

Same as item 1, let's go with a simple solution and improve later.

indyteo commented:
It's curious, I already placed the header name and value fields on the same row
![The two columns](https://user-images.githubusercontent.com/26201395/127869568-bb89baa1-7705-45d9-811d-1272f89a473c.png)
It surely is due to a [bad browser compatibility](https://caniuse.com/css-grid), on which browser did you get this render?

I meant to force the check of the template compile to validate the configuration (i.e. not being able anymore to save an invalid template). Do you think it's a good point to allow users to save invalid configuration, maybe because it's a draft?

I'll fix it so, only the clickable elements will get the pointer cursor :smile:

indyteo commented:
> I think it was Safari. Maybe need to use the HStack instead of Grids?

I first had some issues with HStacks when I tried them, it was why I switched to grids, but I'll try to find a solution with HStacks

> We could validate when saving it as active, but allow invalid template if saved as inactive, what you think?

Excellent idea, it seems to be a good choice, and very simple to do. I like that

> I'll review the code this in and come back with my feedback in case you want to wait and batch changes

I'll do those two small features right now and will adapt the rest later when you'll have reviewed it, no problem!

goenning commented on app/services/webhook/props.go:
```diff
@@ -0,0 +1,161 @@
+	})
+}
+
+func TenantStatus(status int) string {
```
would be better to implement MarshalText/UnmarshalText/String on the enum and use `tenant.Status.String()`

all enums should have this now, but it's possible that I missed some as I added this late

indyteo replied:
I didn't find any method to get a string representation, that's why I did mine, but it's true it would be better to add one instead!

goenning commented on app/services/webhook/webhook.go:
```diff
@@ -0,0 +1,173 @@
+	bus.AddHandler(getWebhookProps)
+}
+
+func triggerWebhook(ctx context.Context, c *cmd.TriggerWebhook) error {
```
It seems like this is only used for testing, maybe we rename it to "testWebhook" or something similar? looking at triggerWebhook it gave me the impression this was the real trigger.

by using a different name here, maybe the one belo could then be just "cmd.TriggerWebhooks"

indyteo replied:
By testing you mean the "Trigger" button in the admin panel I suppose? Maybe `TestWebhook` would be confusing with unit tests, so `DummyTriggerWebhook`. Even if I don't feel very convinced by this name... Because in fact this is really triggering the Webhook, just with dummy props, not only testing the templates and URL validity for example.
I agree about the rename for the other command

goenning replied:
Which makes me think that maybe the button should be "Test", just like the OAUth one? We test a webhook by triggering it with dummy props, but the action itself is still "Test" in my view.

indyteo replied:
I renamed those elements but kept the `WebhookTriggerResult` because it would be used by the hypothetical execution logging system

goenning commented on app/services/webhook/webhook.go:
```diff
@@ -0,0 +1,173 @@
+		"Error":   error,
+	})
+
+	webhooks := &query.MarkWebhookAsFailed{ID: result.Webhook.ID}
```
after a webhook is marked as Failed we never process it again? unless someone re-enable it?

indyteo replied:
Yes, it's the point. An extra task here would be to notify the admins by email, but I didn't find the time to do it, and with the internationalization going, emails were being converted to translated templates at the time and I didn't risked myself into this...
But in theory, when a Webhook has failed, there's a reason behind this and I found this solution instead of keeping failing logs in an extra DB table (but maybe it would be better to do this, and thus being able in the admin panel to see the real failed request instead of a manually triggered one. It would make sense, what are your thought about this?).

goenning replied:
Yeah, it's a lot more complex and we can implement that separately 👍

An email is probably a good starting point, otherwise webhook will get disabled silently. We can add this in a following PR

goenning commented on app/services/webhook/props.go:
```diff
@@ -0,0 +1,161 @@
+	if user == nil {
+		return
+	}
+	role, _ := user.Role.MarshalText()
```
most enums should have a String() function (which comes from the Stringer interface), have a look at the AvatarType enum, doing that means we can inline this on the function below

indyteo replied:
Interesting, I didn't noticed it was on the `AvatarType` enum. I'll implement it on the `Role` enum as well so :)

goenning commented on app/services/webhook/props.go:
```diff
@@ -0,0 +1,161 @@
+	})
+}
+
+func DescribePost(props dto.Props, post *entity.Post, keyPrefix, baseURL string, complete, author bool) {
```
instead of baseURL here, why not get the full url from the task? we'd avoid building the url in multiple places

goenning replied:
I'm finding a bit confusing the use of complete and author params. What do they mean?

indyteo replied:
Concerning the single URL build, I think it would be counter-productive regarding the dummy props generation. Maybe we should add a `Url(Context)` method on `Post` to build the post URL in the given context?

indyteo replied:
And for the params there are whether to include a complete post definition and the post author. I agree the names aren't well chosen, but a `complete` post definition include data that cannot be defined on post creation (votes, tags, response, etc...)

goenning commented on migrations/202107061359_create_webhooks.sql:
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE IF NOT EXISTS webhooks (
+  id SERIAL PRIMARY KEY,
+  name VARCHAR(60) NOT NULL,
```
name is 60 here and 100 on validation. I think 60 is enough, so just align the number there please

indyteo replied:
Damn, that's a lot of inconsistency... I must be tired at the time I did this 😆

goenning commented on migrations/202107061359_create_webhooks.sql:
```diff
@@ -0,0 +1,12 @@
+  name VARCHAR(60) NOT NULL,
+  type SMALLINT NOT NULL,
+  status SMALLINT NOT NULL,
+  url VARCHAR(256) NOT NULL,
```
same for URL, I'd also suggest using 300 to align with the other tables and update the action validation as well

indyteo replied:
I'll change it to a `TEXT` type because it's a template and it may be larger than we expect (with route or GET parameters defined from the Go template). I've adapted the validation to handle the parsing before calling URL validation

goenning commented on migrations/202107061359_create_webhooks.sql:
```diff
@@ -0,0 +1,12 @@
+  status SMALLINT NOT NULL,
+  url VARCHAR(256) NOT NULL,
+  content TEXT NULL,
+  http_method VARCHAR(16) NOT NULL,
```
same

goenning commented on migrations/202107061359_create_webhooks.sql:
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE IF NOT EXISTS webhooks (
```
and this file will have to increment its timestamp to be executed

goenning reviewed (commented):
oh wow, this is a big PR :D

I went through all the backend files, will review frontend later this week.

goenning commented:
Overall looks really good! Happy to see this being registered as a service :) 

I have mixed feelings about the DescribeTenant / DescribePost / DescribeUser but I can't think of a cleaner solution at the moment. The only thing that comes to mind would be a webhook specific props that you'd do:

```
p := webhook.Props{}
p.SetUser(user, ..)
p.SetPost(user, ..)
p.Set("something", ..)
```

Not sure what benefits we'd get, but at least we wouldn't have the public generics Describe* that are meant to be used only on this scenario.

goenning commented on public/assets/styles/utility/display.scss:
```diff
@@ -103,6 +103,10 @@
 }
 
+.justify-full > * {
+  flex: 1 1 0px; // px unit is required to support IE
```
we don't support IE :)

indyteo replied:
Oh, what a good news :D

goenning commented on public/components/common/Loader.tsx:
```diff
@@ -3,12 +3,22 @@ import "./Loader.scss"
-  return show ? <div className="c-loader" /> : null
+  return show ? (
+    <div className={props.className}>
+      <div className="c-loader" />
```
the `.c-{name} is often the top level element returned by the component. Can we add the className to the existing  div?

indyteo replied:
Hmm I need to double check this, there may be a reason why I did it like that, or maybe adapt the Loader's SCSS to target the inner div (maybe a `c-loader__spinner` and a `c-loader__text`?)

goenning commented on public/pages/Administration/components/webhook/WebhookForm.tsx:
```diff
@@ -0,0 +1,286 @@
+  )
+}
+
+export class WebhookForm extends React.Component<WebhookFormProps, WebhookFormState> {
```
I'm slowly migration all older class components into functions+hook. Do you mind converting this one so we don't add new tech debt?

indyteo replied:
Yep, no problem I understand. Good to use those (no so) new React features

goenning commented on public/pages/Administration/components/webhook/WebhookForm.tsx:
```diff
@@ -0,0 +1,286 @@
+            </Button>
+          </HStack>
+        </Form>
+        <WebhookTemplateInfo type={this.state.type} isModalOpen={this.state.isModalOpen} onModalClose={this.hideModal} />
```
rename component to WebhookTemplateInfoModal to follow the convention

indyteo replied:
It will be done

goenning commented on public/pages/Administration/components/webhook/WebhookProperties.scss:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  &__prop {
+    font-family: $font-code;
```
there's a `get("font.family.code")`

indyteo replied:
I searched it at the moment, but didn't spotted it, I'll switch to it

goenning commented on public/pages/Administration/pages/ManageWebhooks.page.tsx:
```diff
@@ -0,0 +1,180 @@
+  )
+}
+
+export default class ManageWebhooksPage extends AdminBasePage<ManageWebhooksPageProps, ManageWebhooksPageState> {
```
same here, if you could cover to function components please (see GeneralSettingsPage as an example to replace class inheritance)

indyteo replied:
I'll check this to help me make this inheritance, I have no idea about what to do it right now :o
By the way, shouldn't we deprecate the elements we're trying to move away? It may avoid theses troubles next time

goenning commented on public/services/actions/webhook.ts:
```diff
@@ -0,0 +1,43 @@
+import { http, Result, StringObject } from "@fider/services"
+import { HttpHeaders, WebhookPreviewResult, WebhookStatus, WebhookTriggerResult, WebhookType } from "@fider/models"
+
+export const createWebhook = async (
```
because of the long list of params here, would it be worth having a type for it?

indyteo replied:
I think it was the reason why I made an extra (maybe already existing) TypeScript type definition, but I kinda messed up with it to be honest. I'll try to improve this little issue

goenning commented on public/services/utils.ts:
```diff
@@ -111,3 +111,11 @@ export const truncate = (input: string, maxLength: number): string => {
   return input
 }
+
+export type Optional<T> = {
```
insn't this the same as `Partial<T>` which comes by default with typescript?

indyteo replied:
Looking at the `Partial<T>` type definition, it seems it is in fact exactly the same. Happy to know I wasn't the only one to do this, I suppose? 😆 (I really didn't know so much about TypeScript, and sometime I find easier to do things by myself, but when it already exist, it's good to use it!)

indyteo commented on app/models/entity/webhook.go:
```diff
@@ -0,0 +1,60 @@
+	Url                   string             `json:"url" db:"url"`
+	Content               string             `json:"content" db:"content"`
+	HttpMethod            string             `json:"http_method" db:"http_method"`
+	AdditionalHttpHeaders HttpHeaders        `json:"additional_http_headers" db:"additional_http_headers"`
```
Initially, I was thinking of a global configuration, then additional per Webhook, but I changed my mind on the way and this name remained unchanged, but you're right I'll rename it it'll be finer :)

indyteo commented on app/services/sqlstore/postgres/webhook.go:
```diff
@@ -0,0 +1,138 @@
+func getWebhook(ctx context.Context, q *query.GetWebhook) error {
+	return using(ctx, func(trx *dbx.Trx, tenant *entity.Tenant, user *entity.User) error {
+		webhook := &entity.Webhook{}
+		if tenant != nil {
```
Hmm yes you're probably right, but this is my trend to always check values when I'm not sure, even if it's not really necessary, I'll remove it

indyteo commented on app/services/sqlstore/postgres/webhook.go:
```diff
@@ -0,0 +1,138 @@
+			err := trx.Select(&webhooks, `
+			SELECT id, name, type, status, url, content, http_method, additional_http_headers 
+			FROM webhooks 
+			WHERE tenant_id = $1 AND type = $2 AND status = 1 
```
Indeed, it'll be cleaner

indyteo commented:
Finally, about the `webhook.Props` it's a really good idea, I like it! It's much more cleaner, I'll adapt the code to do so!

indyteo commented:
I had to do some refactoring to deal with cyclic imports, but I think I went on all your reviews! 😊

goenning commented on app/actions/webhook.go:
```diff
@@ -0,0 +1,137 @@
+	if action.Name == "" {
+		result.AddFieldFailure("name", "Name is required.")
+	} else if len(action.Name) > 60 {
+		result.AddFieldFailure("name", "Name must have less than 100 characters.")
```
60 :)

indyteo replied:
Damn I must be super tired to miss that 😆

goenning reviewed (commented):
We're nearly there! :D 

1. One validation error mentioned above.
2. I've been playing with different parameters for the body and noticed that `post_status` is a string and `post_old_status` is an integer. I'd expect both to be the same type and string is probably better.

On a similar note, it's not possible to quote an integer, so `{{ quote .post_number }}` doesn't work. What if we change quote to force conversion to string? So that `"number": {{ .post_number }}` would result in a number and `"number": {{ quote .post_number }}` would be a string.

I think these two properties add more noise and won't be useful, we could remove it.
- tenant_invitation
- tenant_welcome_message

3. Content outside the modal, do you see the same? Tested on Safari (Mac) and Chrome (Mac).
![image](https://user-images.githubusercontent.com/94755/129409577-81bbb3e5-506e-426c-89e4-42e8b0979901.png)

indyteo commented:
Ok, I fixed the missing post status `.Name()` calls, removed useless tenants props and ensured the `quote` function properly en-quote the value. I also noticed there were no way to just escape a string (without adding extra quotes), so I added the `escape` function (example: `"title": "New post: {{ escape .post_title }}"`)

About the overflowing modal, on Firefox it's clean, I tested it on Safari Mobile (on my iPad), and it doesn't seemed to be overflowing (except when reducing the window to a very small size, so there may be an issue with responsive here, I think I found a workaround). I replaced all the HTML `<table>` because they were horrible when coming to responsive, please have a look now, it should be better :smile:

goenning commented:
Yes, it's a lot better now. The only issue outstanding is long values without space (such as URL). Maybe apply a `word-break: break-all;`. I can't think of anything else.

<img width="922" alt="Screenshot 2021-08-17 at 19 41 54" src="https://user-images.githubusercontent.com/94755/129782643-824531bc-393b-4770-9550-c354cdbe9c15.png">

After that we're good to go :)

indyteo commented:
I'm unable to reproduce the issue on my side (it get wrapped correctly on the next line), could you test if the `word-break` CSS rule do the trick for you?
I used to apply this CSS to break long words on other projects (I know the `hyphens` works well on Firefox but not on other browsers):
```css
overflow-wrap: break-word;
hyphens: auto;
```
So if the `word-break` don't work, maybe try `overflow-wrap`
